### PR TITLE
Notification center and refactor

### LINKED ISSUE #2279: [FR] Notification centre
**Problem**
All the notifications are just email and can get crowded

**Suggested solution**
Some kind of centre where past notifications can be viewed would be great.

**Examples of other systems**
Github would be a great example.

**Do you want to develop this?**
Yes

### LINKED ISSUE #2282: [FR] Multiple notification delivery options per user
**Problem**
Currently only the primary mail is used for notification delivery

**Suggested solution**
- [x] Refactor the current notification system (see #2279) to allow multiple delivery methods
- [x] Give users the option to add notification methods
- [x] Declare notification categories
~~- [ ] Add a new mixin that registers new notification types (and delivery options)~~ -> 2385
~~- [ ] Possibly use the scopes provided by SSO.~~ -> not feasable

**Describe alternatives you've considered**
The new notification system is already good, this would be an addon

**Examples of other systems**
Zabbix, github enterprise, gitlab

**Do you want to develop this?**
Yes

### PR DESCRIPTION
**Pull request by matmair**
This PR refactors the notification system to support multiple delivery methods.
It also adds a notification indicator in the navbar as well as a notification flyout and a notification page.

Closes #2279 
Closes #2282 

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/2372"></a>

### CODE REVIEW

**Comment by matmair** [comment]
grafik

File: InvenTree/common/api.py [hunk 1]
```diff
@@ -130,8 +131,104 @@ class UserSettingsDetail(generics.RetrieveUpdateAPIView):
...
+            user = self.request.user
+        except AttributeError:
+            return common.models.NotificationMessage.objects.none()
+
+        queryset = super().filter_queryset(queryset)
+        queryset = queryset.filter(user=user)
+        return queryset
+
+
+class NotificationDetail(generics.RetrieveDestroyAPIView):
```
Comment (matmair) [inline InvenTree/common/api.py:175]: @SchrodingersGat any idea why the new API views won't work here? I am trying to run this and i always get:
'DELETE action unavailable at /api/notifications/3/'

``` python
constructForm(`/api/notifications/${$(this).attr('pk')}/`, {
        method: 'DELETE',
        title: '{% trans "Delete Notification" %}',
        onSuccess: function(data) {
            updateNotificationTables();
        }
    });
```

Comment (SchrodingersGat) [reply InvenTree/common/api.py:175]: Have you assigned the new `NotificationMessage` model to the permissions groups in `users.models`?

Comment (matmair) [reply InvenTree/common/api.py:175]: Ah I see, it was in RULESET_IGNORE

Comment (matmair) [reply InvenTree/common/api.py:175]: @SchrodingersGat am I allowed to add a new RULESET_MODELS category? I need something that every user is assigned.

Comment (matmair) [reply InvenTree/common/api.py:175]: I thought about 'user' or 'base'

Comment (SchrodingersGat) [reply InvenTree/common/api.py:175]: If it was in `RULESET_IGNORE` then it *should* return "True" for any permission check:

https://github.com/inventree/InvenTree/blob/0cb66a45ae26274bfd711e8ac8017f7a0c6dfc2c/InvenTree/users/models.py#L207

Might need to trace through the code and see why the permissions checks are not running properly.

the API permissions are determined here:

https://github.com/inventree/InvenTree/blob/master/InvenTree/InvenTree/metadata.py

Comment (matmair) [reply InvenTree/common/api.py:175]: @SchrodingersGat I think I found a flaw in the current metadata code that makes it that the metadata inclusion does not work. If a view does not provide a `POST`or `PUT` function `DELETE` and `GET`?  will also not work, you can see this when you create an Endpoint that inherits from `RetrieveDestroyAPIView` as this PR does.
A possible fix could be https://github.com/inventree/InvenTree/pull/2397 but I am not sure if I just misunderstand the functions of the new API forms. It is not the permissions, I used a superuser for all tests now.

**Comment by matmair** [comment]
@SchrodingersGat this would be ready for review and merge

**Comment by SchrodingersGat** [comment]
@matmair nice work, I'll wait until we've split off a release branch for `0.6.x` before looking at this one.

**Comment by matmair** [comment]
@SchrodingersGat please do not merge till #2621 #2620 #2564 are merged

**Comment by matmair** [comment]
@SchrodingersGat this would be ready for a review now.

This PR adds a notification view (like GitHub but fewer functions) and remasters the way notifications are sent -> getting it ready for multiple delivery methods and plugins that send messages.

**Comment by SchrodingersGat** [comment]
@matmair wonderful, I'll try to find some time soon, but it may be a couple of weeks! Very busy rn

**Comment by matmair** [comment]
@SchrodingersGat dont stress about it!

**Comment by matmair** [comment]
@schrodingersGat any timeline to reviewing this?
Currently considering to either work on this + plugin issues or ordering / pricing stuff.

**Comment by SchrodingersGat** [comment]
@matmair hopefully in the next couple of weeks :)

**Comment by SchrodingersGat** [comment]
@matmair looks like there is an issue with an untranslated string:

**Comment by SchrodingersGat** [comment]
Icons:

image

"Notifications" - should be the bell icon
"Inbox" - envelope icon
"history" - clock / history icon

**Comment by SchrodingersGat** [comment]
It would be useful to have a "mark all as read" button 

image

File: InvenTree/InvenTree/static/script/inventree/inventree.js [hunk 2]
```diff
@@ -216,8 +217,25 @@ function inventreeDocReady() {
...
+    $(document).focus(function(){
+        startNotificationWatcher();
+    });
+
+    // kill notification watcher if focus is lost -> respect your users cycles
+    $(document).blur(function(){
+        stopNotificationWatcher();
+    });
+
+    $('#offcanvasRight').on('show.bs.offcanvas', openNotificationPanel)  // listener for opening the notification panel
```
Comment (SchrodingersGat) [inline InvenTree/InvenTree/static/script/inventree/inventree.js:234]: Missing semicolon

File: InvenTree/InvenTree/static/script/inventree/inventree.js [hunk 3]
```diff
@@ -216,8 +217,25 @@ function inventreeDocReady() {
...
+    $('#offcanvasRight').on('hidden.bs.offcanvas', closeNotificationPanel)  // listener for closing the notification panel
```
Comment (SchrodingersGat) [inline InvenTree/InvenTree/static/script/inventree/inventree.js:235]: Missing semicolon

File: InvenTree/InvenTree/views.py [hunk 4]
```diff
@@ -904,3 +904,18 @@ def get_context_data(self, **kwargs):
...
+    """ View for showing notifications
+    """
+
+    template_name = "InvenTree/notifications/notifications.html"
+
+    def get_context_data(self, **kwargs):
+
+        ctx = super().get_context_data(**kwargs).copy()
+
+        # ctx['settings'] = InvenTreeSetting.objects.all().order_by('key')
```
Comment (SchrodingersGat) [inline InvenTree/InvenTree/views.py:919]: does get_context_data need to be reimplemented here, now that this line is commented out?

File: InvenTree/templates/InvenTree/notifications/history.html [hunk 5]
```diff
@@ -0,0 +1,25 @@
+{% extends "panel.html" %}
+
+{% load i18n %}
+{% load inventree_extras %}
+
+{% block label %}history{% endblock %}
+
+{% block heading %}
+{% trans "History" %}
```
Comment (SchrodingersGat) [inline InvenTree/templates/InvenTree/notifications/history.html:9]: A more descriptive text e.g. "Notification History" would be useful

File: InvenTree/templates/InvenTree/notifications/inbox.html [hunk 6]
```diff
@@ -0,0 +1,25 @@
+{% extends "panel.html" %}
+
+{% load i18n %}
+{% load inventree_extras %}
+
+{% block label %}inbox{% endblock %}
+
+{% block heading %}
+{% trans "Inbox" %}
```
Comment (SchrodingersGat) [inline InvenTree/templates/InvenTree/notifications/inbox.html:9]: A more descriptive text e.g. "Pending Notifications" would be useful

**Comment by matmair** [comment]
@SchrodingersGat I added all requested fixes / changes

**Comment by SchrodingersGat** [comment]
Thanks for the changes. I'll take another look. btw this is looking *really* great, I love the way you have implemented this.

**Review by SchrodingersGat — APPROVED** [review]

**Comment by SchrodingersGat** [comment]
Thanks @matmair a really great addition. If you have some time for documentation before the next release, that would be great.

**Comment by matmair** [comment]
> Thanks for the changes. I'll take another look. btw this is looking *really* great, I love the way you have implemented this.

Thanks. It should be rather easy to write a lot of different notification methods (classes) with the least possible code and duplication.

I would have liked a computational less expensive method to check for new notifications like Websockets with django channels. But that would add a lot of possible setup/install problems.

**Pull request merged**